host/sgx/loadelf.c

```diff
@@ -695,6 +715,171 @@ static oe_result_t _set_uint64_t_dynamic_symbol_value(
     return result;
 }
 
+static oe_result_t _link_elf_image(
+    oe_enclave_elf_image_t* image,
+    oe_enclave_elf_image_t* dependency)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    assert(image);
+    assert(dependency);
+
+    const elf64_sym_t* symtab = NULL;
+    size_t symtab_size = 0;
+    if (elf64_get_dynamic_symbol_table(&image->elf, &symtab, &symtab_size) != 0)
+        goto done;
+
+    /* Iterate through relocation records in the target image */
+    elf64_rela_t* relocs = (elf64_rela_t*)image->reloc_data;
+    uint64_t nrelocs = image->reloc_size / sizeof(relocs[0]);
+
+    for (size_t i = 0; i < nrelocs; i++)
+    {
+        elf64_rela_t* p = &relocs[i];
+
+        /* Skip the zero-padded bytes */
+        if (p->r_offset == 0)
+            break;
+
+        /* Fix up the r_offset based on the image_rva */
+        p->r_offset += image->image_rva;
+
+        uint64_t reloc_type = ELF64_R_TYPE(p->r_info);
+
+        /* Patch symbolic relocation records to avoid having symbol lookup
+         * in the enclave */
+        if (reloc_type == R_X86_64_GLOB_DAT ||
+            reloc_type == R_X86_64_JUMP_SLOT || reloc_type == R_X86_64_64)
+        {
+            uint64_t symbol_index = ELF64_R_SYM(p->r_info);
+            const elf64_sym_t* symbol = &symtab[symbol_index];
+            const char* name =
+                elf64_get_string_from_dynstr(&image->elf, symbol->st_name);
+            if (name == NULL)
+                OE_RAISE(OE_NOT_FOUND);
+
+            int64_t addend = (reloc_type == R_X86_64_64) ? p->r_addend : 0;
+
+            /* To simplify the in-enclave relocation handling, we convert
+             * all the symbolic relocation types to X86_64_RELATIVE. */
+            p->r_info = (symbol_index << 32) | R_X86_64_RELATIVE;
+
+            /* Find the definition of the symbol in the image itself */
+            elf64_sym_t symbol_definition = {0};
+            if (elf64_find_dynamic_symbol_by_name(
+                    &image->elf, name, &symbol_definition) == 0 &&
+                symbol_definition.st_shndx != SHN_UNDEF)
+            {
+                p->r_addend =
+                    (int64_t)(image->image_rva + symbol_definition.st_value) +
+                    addend;
+            }
+            /* Find the definition of the symbol in the dependent image */
+            else if (
+                elf64_find_dynamic_symbol_by_name(
+                    &dependency->elf, name, &symbol_definition) == 0 &&
+                symbol_definition.st_shndx != SHN_UNDEF)
+            {
+                p->r_addend =
+                    (int64_t)(
+                        dependency->image_rva + symbol_definition.st_value) +
+                    addend;
+            }
+            else
+            {
+                OE_TRACE_WARNING("symbol %s not found\n", name);
```

Comment:
This should be a hard failure in your 2 binary model, no?

Reply:
This allows weak symbol to be unresolved. For example
```
__attribute__((weak)) int sub(int a, int b);
```

Reply:
This allows for unresolved weak symbols. For example
```
__attribute__((weak)) int sub(int a, int b);
```

Reply:
I think you should be able to determine unresolved weak symbols (unused) at static link time and filter for those (they should have a zero value) instead of trying to find the symbol and allowing any unresolved symbol to be only a warning. Weak symbols should only be relevant at static link time, should be marked by the zero value if unused by runtime binding. 

For comparison, I would check that this code behaves the same when given a non-weak symbol that was dynamically linked with `LD_BIND_NOW` enabled in a non-enclave app.

Reply:
Now forcing the error is the unresolved symbol not a weak symbol.